From 0e2c2a21d1e54793752738716ae43ef1759cf625 Mon Sep 17 00:00:00 2001
From: Lei Wen <leiwen@marvell.com>
Date: Mon, 29 Jun 2009 14:32:44 +0800
Subject: [PATCH] fix netconsole hang bug when edit the command

Netcat utility likes to exit when it receives an empty packet,
as it thinks this means EOF. This can easily occur when working
with command line editting with backspace, or tab completion
behavior. Since local tool ncb doesn't have this feature, so it
could workaround netconsole cannot editting command issue.

Signed-off-by: Lei Wen <leiwen@marvell.com>
---
 tools/Makefile   |    4 ++--
 tools/netconsole |   18 +++++++++++++++---
 2 files changed, 17 insertions(+), 5 deletions(-)

diff --git a/tools/Makefile b/tools/Makefile
index 9e9ee15..7362ce2 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -21,10 +21,10 @@
 # MA 02111-1307 USA
 #
 
-BIN_FILES	= img2srec$(SFX) mkimage$(SFX) envcrc$(SFX) ubsha1$(SFX) gen_eth_addr$(SFX) bmp_logo$(SFX)
+BIN_FILES	= img2srec$(SFX) mkimage$(SFX) envcrc$(SFX) ubsha1$(SFX) gen_eth_addr$(SFX) bmp_logo$(SFX) ncb$(SFX)
 
 OBJ_LINKS	= env_embedded.o crc32.o md5.o sha1.o image.o
-OBJ_FILES	= img2srec.o mkimage.o envcrc.o ubsha1.o gen_eth_addr.o bmp_logo.o
+OBJ_FILES	= img2srec.o mkimage.o envcrc.o ubsha1.o gen_eth_addr.o bmp_logo.o ncb.o
 
 ifeq ($(ARCH),mips)
 BIN_FILES	+= inca-swap-bytes$(SFX)
diff --git a/tools/netconsole b/tools/netconsole
index 09c8981..aa30d9b 100755
--- a/tools/netconsole
+++ b/tools/netconsole
@@ -31,12 +31,24 @@ if [ -z "${ip}" ] || [ -n "$3" ] ; then
 fi
 
 for nc in netcat nc ; do
-	type ${nc} >/dev/null && break
+	type ${nc} >/dev/null 2>&1 && break
 done
 
 trap "stty icanon echo intr ^C" 0 2 3 5 10 13 15
 echo "NOTE: the interrupt signal (normally ^C) has been remapped to ^T"
 
 stty -icanon -echo intr ^T
-${nc} -u -l -p ${port} < /dev/null &
-exec ${nc} -u ${ip} ${port}
+(
+if [ -x ./ncb ] ; then
+	exec ./ncb ${port}
+elif [ -x ${0%/*}/ncb ] ; then
+	exec ${0%/*}/ncb ${port}
+else
+	while ${nc} -u -l -p ${port} < /dev/null ; do
+		:
+	done
+fi
+) &
+pid=$!
+${nc} -u ${ip} ${port}
+kill $pid 2>/dev/null
-- 
1.6.0.4

